### ISSUE TITLE
Add backtraces to errors

### ISSUE THREAD

**Issue by jedel1043** [boa-dev/boa#3726]
As part of our focus on improving our Error APIs, we need to introduce backtraces to errors generated at execution. This includes:

- [ ] Adding a `Backtrace` struct.
- [ ] Retrofitting our VM to hold the current backtrace.
- [ ] Adding names to all native functions.
- [ ] Storing the current backtrace inside every thrown `JsError`.

**Comment by raskad** [comment]
Do you think it makes sense to consider TCO when we implement this? We could make sure both work together or decide that we do not want to implement TCO.

See the discussion in #3510.

**Comment by jedel1043** [comment]
Yep. Though, I think a well designed backtrace API would isolate both issues e.g. errors would get a stacktrace without knowing the internals of whether the VM has a "phantom trace" or not.
